core/engine/src/vm/opcode/args.rs

```diff
@@ -286,48 +287,177 @@ impl Argument for (VaryingOperand, i16) {
+                ((arg1.into(), arg2), pos + 5)
+            }
+            Format::U16 => {
+                assert!(bytes.len() >= pos + 6, "buffer too small to read arguments");
+                let (arg1, arg2) = unsafe {
+                    (
+                        read_unchecked::<u16>(bytes, pos),
+                        read_unchecked::<i32>(bytes, pos + 2),
+                    )
+                };
+                ((arg1.into(), arg2), pos + 6)
+            }
+            Format::U32 => {
+                let ((arg1, arg2), pos) = read::<(u32, i32)>(bytes, pos);
+                ((arg1.into(), arg2), pos)
+            }
+        }
     }
 }
 
 impl Argument for (VaryingOperand, f32) {
     fn encode(self, bytes: &mut Vec<u8>) {
-        write_u32(bytes, self.0.value);
-        write_f32(bytes, self.1);
+        match self.0.variant() {
+            VaryingOperandVariant::U8(value) => {
+                write_format(bytes, Format::U8);
+                write_u8(bytes, value);
+                write_f32(bytes,self.1);
+            }
+            VaryingOperandVariant::U16(value) => {
+                write_format(bytes, Format::U16);
+                write_u16(bytes, value);
+                write_f32(bytes,self.1);
+            }
+            VaryingOperandVariant::U32(value) => {
+                write_format(bytes, Format::U32);
+                write_u32(bytes, value);
+                write_f32(bytes,self.1);
+            }
+        }
     }
 
     fn decode(bytes: &[u8], pos: usize) -> (Self, usize) {
-        let ((arg1, arg2), pos) = read::<(u32, u32)>(bytes, pos);
-        ((arg1.into(), f32::from_bits(arg2)), pos)
+
+        let format = Format::from(bytes[pos]);
+        let pos = pos + 1;
+
+        match format {
+            Format::U8 => {
+                assert!(bytes.len() >= pos + 5, "buffer too small to read arguments");
+                let (arg1, arg2) = unsafe {
+                    (
+                        read_unchecked::<u8>(bytes, pos),
+                        read_unchecked::<f32>(bytes, pos + 1),
+                    )
+                };
+                ((arg1.into(), arg2), pos + 5)
+            }
+            Format::U16 => {
+                assert!(bytes.len() >= pos + 6, "buffer too small to read arguments");
+                let (arg1, arg2) = unsafe {
+                    (
+                        read_unchecked::<u16>(bytes, pos),
+                        read_unchecked::<f32>(bytes, pos + 2),
+                    )
+                };
+                ((arg1.into(), arg2), pos + 6)
+            }
+            Format::U32 => {
+                assert!(bytes.len() >= pos + 8, "buffer too small to read arguments");
+                let (arg1, arg2) = unsafe {
+                    (
+                        read_unchecked::<u32>(bytes, pos),
+                        read_unchecked::<f32>(bytes, pos + 4),
+                    )
+                };
+                ((arg1.into(), arg2), pos + 8)
```

Comment:
Couldn't we use `read` here?
```suggestion
                let ((arg1, arg2), pos) = read::<(u32, f32)>(bytes, pos);
                ((arg1.into(), arg2), pos)
```